### PR TITLE
handle exit and avoid threads hanging

### PR DESCRIPTION
**Pull request by rphmeier**

### CODE REVIEW

File: polkadot/consensus/src/service.rs [hunk 1]
```diff
@@ -265,9 +265,11 @@ impl Service {
 		client: Arc<C>,
 		network: Arc<net::ConsensusService>,
 		transaction_pool: Arc<Mutex<TransactionPool>>,
-		key: ed25519::Pair
+		key: ed25519::Pair,
+		exit: ::exit_future::Exit,
```
Comment (arkpar) [inline polkadot/consensus/src/service.rs]: As I've mentioned in personal discussion I can't say I like this. This just introduced additional complexity to the service interface and leaks away implementation details. What happens if the service is dropped before `exit` is signalled? This is something the user of this interface should not worry about.

File: polkadot/service/src/lib.rs [hunk 2]
```diff
@@ -284,40 +291,58 @@ impl Service {
...
+					debug!("Polkadot service event loop shutdown with {:?}", e);
+				}
+				debug!("Polkadot service shutdown");
+			})
+		};
+
+		// before returning, make sure the network is started. avoids a race
+		// between the drop killing notification listeners and the new notification
+		// stream being started.
+		barrier.wait();
```
Comment (arkpar) [inline polkadot/service/src/lib.rs]: Is this really necessary? There's no event loop blocked on the network anymore.

Comment (rphmeier) [reply polkadot/service/src/lib.rs]: it's also nice to make sure the network is started before we begin the consensus process.

File: polkadot/service/src/lib.rs [hunk 3]
```diff
@@ -284,7 +291,39 @@ impl Service {
...
+				let events = client.import_notification_stream().for_each(move |notification| {
+					network.on_block_imported(notification.hash, &notification.header);
+					prune_imported(&*client, &*txpool, notification.hash);
+
+					Ok(())
+				});
+
+				core.handle().spawn(events);
+				if let Err(e) = core.run(exit) {
+					debug!("Polkadot service event loop shutdown with {:?}", e);
```
Comment (gavofyork) [inline polkadot/service/src/lib.rs]: not needed fro this PR, but in general i'd like to get away from anonymous messages at anything more verbose than the `info!` level. enabling `debug` and `trace` is a currently a mess, so doesn't really make sense to enable them without specifying a target filter.

Comment (arkpar) [reply polkadot/service/src/lib.rs]: You can always filter by module. I usually use an explicit target to logically group logs coming from several modules under a single filter. If there is just a single module there's no need for that.

**Comment by arkpar** [comment]
`Client::stop_notifications` should be removed

File: polkadot/service/src/lib.rs (hunk 3 above)
Comment (rphmeier) [reply polkadot/service/src/lib.rs]: agree; this was actually just moved from the previous version

**Pull request merged**